src/branch_and_prune.jl

```diff
@@ -5,42 +5,98 @@ export branch_and_prune, Bisection, Newton
 
 diam(x::Root) = diam(x.interval)
 
-
 Base.size(x::Interval) = (1,)
 
 isinterior{N}(X::IntervalBox{N}, Y::IntervalBox{N}) = all(isinterior.(X, Y))
 
+# contractors:
+"""
+    Contractor{F}
 
+    Abstract type for contractors.
 """
-    branch_and_prune(f, X, contractor, tol=1e-3)
+abstract type Contractor{F} end
 
-Generic branch and prune routine for finding isolated roots of a function
-`f:R^n → R^n` in a box.
+export Bisection, Newton
 
-Inputs:
-- `f`: function whose roots will be found
-- `X`: `Interval` or `IntervalBox`
-- `contractor`: function that, when applied to the function `f`, determines
-    the status of a given box `X`. It returns the new box and a symbol indicating
-    the status. Current possible values are `Bisection` and `Newton`.
+struct Bisection{F} <: Contractor{F}
+    f::F
+end
 
-"""
-function branch_and_prune(f, X, contractor, tol=1e-3)
+Bisection{n}(::Type{Val{n}}, f) = Bisection(f)
+
+function (contractor::Bisection)(X)
+    image = contractor.f(X)
+
+    if !(contains_zero(image))
+        return :empty, X
+    end
+
+    return :unknown, X
+end
+
+
+struct Newton{F,FP,O} <: Contractor{F}
+    f::F
+    fp::FP
+    op::O
+end
+
+function Newton(::Type{Val{1}}, f::Function)
```

Comment:
I know this was in my original code, but it seems nasty to have the `Val{1}` stuff.
I think this was basically to get around the fact that you can't extract the dimensions from the function `f`.

Maybe we should have some kind of wrapper for the function that contains the information on its dimensions.

Reply:
If we decide that the `roots` function (rather than the `Newton` one) is responsible to provide derivative, even when none was explicitly given, the dimension is no longer useful as it is only used to choose between `ForwardDiff.derivative` and `ForwardDiff.jacobian`.

Reply:
It could be done by specializing the `roots` function even more:

    function roots(f, X::Interval{T}, contractor::Newton, tol::Float64=1e-3) where {T}
        contract = contractor(f, x -> ForwardDiff.derivative(f, x)))
        branch_and_prune(X, contract, tol)
    end

    function roots(f, X::IntervalBox{T}, contractor::Newton, tol::Float64=1e-3) where {T}
        contract = contractor(f, x -> ForwardDiff.jacobian(f, x)))
        branch_and_prune(X, contract, tol)
    end